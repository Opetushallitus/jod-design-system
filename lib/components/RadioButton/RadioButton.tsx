import { Radio } from '@headlessui/react';

export interface RadioButtonProps {
  /** Text for the component */
  label: string;
  /** Value for the component */
  value: string;
  /** CSS classes for custom styles */
  className?: string;
  /** Component variant */
  variant?: 'default' | 'bordered';
}

export const RadioButton = ({ label, value, className, variant = 'default' }: RadioButtonProps) => {
  return (
    <Radio value={value} className={`${className ? className : ''} flex h-7`.trim()}>
      {({ checked }) => (
        <div className="ds-flex-start ds-flex ds-space-x-4">
          {checked ? <CheckedIcon /> : <UncheckedIcon variant={variant} />}
          <span className="ds-flex ds-items-center ds-text-heading-4 ds-text-black hover:ds-text-accent hover:ds-underline ds-hyphens-auto">
            {label}
          </span>
        </div>
      )}
    </Radio>
  );
};

const CheckedIcon = () => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="16"
      height="16"
      viewBox="0 0 16 16"
      fill="none"
      className="ds-self-center"
    >
      <circle cx="8" cy="8" r="8" className="ds-fill-accent" />
    </svg>
  );
};

const UncheckedIcon = ({ variant }: { variant?: RadioButtonProps['variant'] }) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="16"
      height="16"
      viewBox="0 0 16 16"
      fill="none"
      className="ds-self-center"
    >
      <circle cx="8" cy="8" r="8" className="ds-fill-white" />
      {variant === 'bordered' && <circle cx="8" cy="8" r="7" className="ds-stroke-accent" />}
    </svg>
  );
};
