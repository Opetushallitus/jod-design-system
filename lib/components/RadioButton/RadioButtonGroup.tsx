import React from 'react';
import { Label, RadioGroup } from '@headlessui/react';

export interface RadioButtonGroupProps {
  /** Text for the component */
  label: string;
  /** Value that is currently selected */
  value: string;
  /** Callback function when the value changes */
  onChange: (newValue: string) => void;
  /** Use RadioButton components as children */
  children: React.ReactNode;
  /** CSS classes for custom styles */
  className?: string;
}

export const RadioButtonGroup = ({ label, value, onChange, children, className }: RadioButtonGroupProps) => {
  return (
    <RadioGroup
      value={value}
      onChange={onChange}
      className={`${className ? className : ''} flex flex-col space-y-2`.trim()}
    >
      <Label className="mb-5 text-body-xs text-secondary-gray">{label}</Label>
      {children}
    </RadioGroup>
  );
};
