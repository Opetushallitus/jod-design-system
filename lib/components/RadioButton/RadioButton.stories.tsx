import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import { RadioButton } from './RadioButton';
import { RadioButtonGroup, RadioButtonGroupProps } from './RadioButtonGroup';

const meta = {
  title: 'Forms/RadioButton',
  component: RadioButtonGroup,
  tags: ['autodocs'],
} satisfies Meta<typeof RadioButtonGroup>;

export default meta;

type Story = StoryObj<typeof meta>;

const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?type=design&node-id=746-43288',
  },
};

const DefaultStoryRender = (props: RadioButtonGroupProps) => {
  const [value, setValue] = React.useState<string>('a');
  return <RadioButtonGroup {...props} value={value} onChange={setValue} />;
};
export const Default: Story = {
  parameters: {
    ...parameters,
  },
  args: {
    label: 'Valitse, miten haluat ryhmitellä osaamisesi.',
    value: 'a',
    onChange: fn(),
    children: (
      <>
        <RadioButton label="Lähteiden mukaan" value="a" />
        <RadioButton label="Teemoittain" value="b" />
        <RadioButton label="Aakkosellisesti" value="c" />
      </>
    ),
  },
  render: (args) => {
    return <DefaultStoryRender {...args} />;
  },
};
