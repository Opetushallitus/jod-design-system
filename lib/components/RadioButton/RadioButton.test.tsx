import { render, screen } from '@testing-library/react';
import { describe, expect, it, test, vi } from 'vitest';
import type { ReactNode } from 'react';

import { RadioButtonGroup } from './RadioButtonGroup';
import { RadioButton } from './RadioButton';

const Wrapper = ({ children }: { children: ReactNode }) => {
  return (
    <RadioButtonGroup label="Label" value="a" onChange={vi.fn()}>
      {children}
    </RadioButtonGroup>
  );
};

describe('Snapshot testing', () => {
  test('Default', () => {
    const { container } = render(
      <Wrapper>
        <RadioButton label="A" value="a" />
      </Wrapper>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('RadioButton', () => {
  it('renders the label correctly', () => {
    render(
      <Wrapper>
        <RadioButton label="A" value="a" />
      </Wrapper>,
    );
    expect(screen.getByText('A')).toBeInTheDocument();
  });
});
