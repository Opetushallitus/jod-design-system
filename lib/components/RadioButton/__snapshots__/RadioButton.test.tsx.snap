// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > Default 1`] = `
<div
  aria-labelledby="headlessui-label-:r1:"
  class="flex flex-col space-y-2"
  id="headlessui-radiogroup-:r0:"
  role="radiogroup"
>
  <label
    class="mb-5 text-body-xs text-secondary-gray"
    id="headlessui-label-:r1:"
    role="none"
  >
    Label
  </label>
  <div
    aria-checked="true"
    class="flex h-7"
    data-headlessui-state="checked"
    id="headlessui-radiogroup-option-:r2:"
    role="radio"
    tabindex="0"
  >
    <div
      class="flex-start flex space-x-4"
    >
      <svg
        class="self-center"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          class="stroke-accent"
          cx="12.0352"
          cy="12.0389"
          r="8"
          stroke-width="8"
        />
      </svg>
      <span
        class="flex items-center text-button-md text-primary-gray hover:text-accent hover:underline"
      >
        A
      </span>
    </div>
  </div>
</div>
`;
